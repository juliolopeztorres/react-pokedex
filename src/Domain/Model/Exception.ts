
export default class Exception {

  public static create(message: string, context?: object): Exception {
    return new Exception(message, context)
  }

  private constructor(
    public message: string,
    public context?: object
  ) {
  }
}
