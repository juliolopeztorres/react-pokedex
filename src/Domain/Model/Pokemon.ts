export default class Pokemon {

  constructor(
    public id : number,
    public name : string
  ) {}
}
