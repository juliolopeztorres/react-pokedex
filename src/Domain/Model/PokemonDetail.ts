import BaseStat from "./BaseStat";

export default class PokemonDetail {

  constructor(
    public id : number,
    public name : string,
    public image : string,
    public types : string[],
    public backSprites : string[],
    public baseStats : BaseStat[]
  ) {}
}
