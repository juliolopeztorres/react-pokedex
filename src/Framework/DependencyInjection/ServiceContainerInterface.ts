
export type ServiceType = 'GetPokemonsUseCase' | 'PokemonRepository' | 'LocalStorage' | string

export default interface ServiceContainerInterface {
  getService(name: ServiceType): any;
}
