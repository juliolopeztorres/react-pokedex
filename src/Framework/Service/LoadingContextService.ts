import { createContext } from 'react'

const LoadingContextService = createContext<[boolean, (status: boolean) => void]>([false, () => {}])

export default LoadingContextService
