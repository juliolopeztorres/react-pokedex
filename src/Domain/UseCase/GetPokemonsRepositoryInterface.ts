import Pokemon from "../Model/Pokemon";
import PokemonDetail from "../Model/PokemonDetail";
import { Generation } from "../Model/GenerationInfo";

export default interface GetPokemonsRepositoryInterface {
  getByGeneration(generation: Generation): Promise<Pokemon[]>

  getById(id: number): Promise<PokemonDetail>
}
