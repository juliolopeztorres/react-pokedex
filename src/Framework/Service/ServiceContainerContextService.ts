import { createContext } from 'react'
import ServiceContainerInterface from "../DependencyInjection/ServiceContainerInterface";
import ServiceContainer from "../DependencyInjection/ServiceContainer";

const ServiceContainerContextService = createContext<ServiceContainerInterface>(new ServiceContainer())

export default ServiceContainerContextService
