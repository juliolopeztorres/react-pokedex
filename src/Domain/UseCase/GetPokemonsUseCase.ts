import Exception from "../Model/Exception";
import GetPokemonsRepositoryInterface from "./GetPokemonsRepositoryInterface";
import Pokemon from "../Model/Pokemon";
import PokemonDetail from "../Model/PokemonDetail";
import { Generation } from "../Model/GenerationInfo";

export default class GetPokemonsUseCase {
  constructor(private repository: GetPokemonsRepositoryInterface) {}

  public getAll(generation: Generation): Promise<Pokemon[]>
  {
    return this.repository.getByGeneration(generation)
      .catch((error) => Promise.reject(Exception.create(`Could not recover pokemons for generation ${generation}`, error)))
  }

  public getById(id: number): Promise<PokemonDetail>
  {
    return this.repository.getById(id)
      .catch((error) => Promise.reject(Exception.create(`Could not recover pokemon detail for id ${id}`, error)))
  }
}
