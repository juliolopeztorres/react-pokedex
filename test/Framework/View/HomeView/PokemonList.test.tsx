import React from 'react'
import { createRoot, Root } from "react-dom/client"
import { act } from "@testing-library/react";
import ServiceContainerInterface, { ServiceType } from "../../../../src/Framework/DependencyInjection/ServiceContainerInterface";
import GetPokemonsUseCase from "../../../../src/Domain/UseCase/GetPokemonsUseCase";
import GetPokemonsRepositoryInterface from "../../../../src/Domain/UseCase/GetPokemonsRepositoryInterface";
import Pokemon from "../../../../src/Domain/Model/Pokemon";
import PokemonDetail from "../../../../src/Domain/Model/PokemonDetail";
import Exception from "../../../../src/Domain/Model/Exception";

const logFunctionMock = jest.fn()
jest.mock('../../../../src/Domain/Util/logWithLevel', () => {
  return {
    __esModule: true,
    Level: jest.requireActual('../../../../src/Domain/Util/logWithLevel').Level,
    default: logFunctionMock,
  }
})

import LoadingContextService from "../../../../src/Framework/Service/LoadingContextService";
import ServiceContainerContextService from "../../../../src/Framework/Service/ServiceContainerContextService";
import PokemonList from "../../../../src/Framework/View/HomeView/PokemonList";

let root: Root
let container: HTMLElement
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div")
  document.body.appendChild(container)

  root = createRoot(container)
})

afterEach(() => {
  // cleanup on exiting
  act(() => root.unmount())

  container.remove()
})

it('should render', async () => {
  const setLoading = jest.fn()
  const onPokemonClicked = jest.fn()

  await act(async () => {
    root.render(
      <ServiceContainerContextService.Provider value={new class implements ServiceContainerInterface {
        getService(name : ServiceType) : any {
          if (name === 'GetPokemonsUseCase') {
            return new GetPokemonsUseCase(new class implements GetPokemonsRepositoryInterface {
              getByGeneration(generation : string) : Promise<Pokemon[]> {
                return Promise.resolve([{id: 1, name: 'bulbasaur'}, {id: 2, name: 'charizard'}])
              }

              getById(id : number) : Promise<PokemonDetail> {
                throw new Error('Not implemented')
              }
            })
          }

          throw new Error('Unexpected service requested')
        }
      }}>
        <LoadingContextService.Provider value={[false, setLoading]}>
          <PokemonList
            searchPokemonName={'bulba'}
            onPokemonClicked={onPokemonClicked}
            generation={'1st'}
          />
        </LoadingContextService.Provider>
      </ServiceContainerContextService.Provider>
    )
  })

  expect(setLoading).toHaveBeenCalledTimes(2)
  expect(setLoading).toHaveBeenNthCalledWith(1, true)
  expect(setLoading).toHaveBeenNthCalledWith(2, false)

  const list = container.querySelector('ul[data-test-id="list"]')
  expect(list).toBeTruthy()

  expect(list!.childElementCount).toBe(1)
  const listElement = list!.firstElementChild!

  expect(listElement.textContent).toContain('1. Bulbasaur')

  expect(onPokemonClicked).not.toHaveBeenCalled()
  act(() => {
    listElement.dispatchEvent(new MouseEvent('click', {bubbles: true}))
  })

  expect(onPokemonClicked).toHaveBeenCalledTimes(1)
  expect(onPokemonClicked).toHaveBeenLastCalledWith(new Pokemon(1, 'bulbasaur'))

  expect(container).toMatchSnapshot()
})

it('should log if use case fails', async () => {
  const setLoading = jest.fn()
  const onPokemonClicked = jest.fn()

  const expectedException = Exception.create('Error coming from Repository')

  await act(async () => {
    root.render(
      <ServiceContainerContextService.Provider value={new class implements ServiceContainerInterface {
        getService(name : ServiceType) : any {
          if (name === 'GetPokemonsUseCase') {
            return new GetPokemonsUseCase(new class implements GetPokemonsRepositoryInterface {
              getByGeneration(generation : string) : Promise<Pokemon[]> {
                return Promise.reject(expectedException)
              }

              getById(id : number) : Promise<PokemonDetail> {
                throw new Error('Not implemented')
              }
            })
          }

          throw new Error('Unexpected service requested')
        }
      }}>
        <LoadingContextService.Provider value={[false, setLoading]}>
          <PokemonList
            searchPokemonName={'bulba'}
            onPokemonClicked={onPokemonClicked}
            generation={'1st'}
          />
        </LoadingContextService.Provider>
      </ServiceContainerContextService.Provider>
    )
  })

  expect(setLoading).toHaveBeenCalledTimes(2)
  expect(setLoading).toHaveBeenNthCalledWith(1, true)
  expect(setLoading).toHaveBeenNthCalledWith(2, false)

  expect(logFunctionMock).toHaveBeenCalledTimes(1)
  expect(logFunctionMock).toHaveBeenLastCalledWith(
    'ðŸ”´',
    'error getting pokemons from repository',
    { indentation: 2, context: Exception.create(
      'Could not recover pokemons for generation 1st',
        expectedException
      )}
  )

  expect(container).toMatchSnapshot()
})

it('should render correctly spaces ahead pokemon numbers', async () => {
  const setLoading = jest.fn()
  const onPokemonClicked = jest.fn()

  await act(async () => {
    root.render(
      <ServiceContainerContextService.Provider value={new class implements ServiceContainerInterface {
        getService(name : ServiceType) : any {
          if (name === 'GetPokemonsUseCase') {
            return new GetPokemonsUseCase(new class implements GetPokemonsRepositoryInterface {
              getByGeneration(generation : string) : Promise<Pokemon[]> {
                return Promise.resolve([{id: 1, name: 'bulbasaur'}, {id: 20, name: 'charizard'}, {id: 300, name: 'squirtle'}])
              }

              getById(id : number) : Promise<PokemonDetail> {
                throw new Error('Not implemented')
              }
            })
          }

          throw new Error('Unexpected service requested')
        }
      }}>
        <LoadingContextService.Provider value={[false, setLoading]}>
          <PokemonList
            searchPokemonName={''}
            onPokemonClicked={onPokemonClicked}
            generation={'1st'}
          />
        </LoadingContextService.Provider>
      </ServiceContainerContextService.Provider>
    )
  })

  expect(setLoading).toHaveBeenCalledTimes(2)
  expect(setLoading).toHaveBeenNthCalledWith(1, true)
  expect(setLoading).toHaveBeenNthCalledWith(2, false)

  const list = container.querySelector('ul[data-test-id="list"]')
  expect(list).toBeTruthy()

  expect(list!.childElementCount).toBe(3)

  expect(list!.firstElementChild!.textContent).toEqual('Â Â 1. Bulbasaur')
  expect(list!.children[1].textContent).toEqual('Â 20. Charizard')
  expect(list!.children[2].textContent).toEqual('300. Squirtle')

  expect(container).toMatchSnapshot()
})
