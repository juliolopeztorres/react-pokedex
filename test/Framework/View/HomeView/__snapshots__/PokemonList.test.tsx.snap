// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <ul
    class="
        columns-2
        md:columns-3
        lg:columns-4
        xl:columns-5
        font-mono
        text-xs
        md:text-sm
        "
    data-test-id="list"
  >
    <li
      class="
            text-slate-800
            dark:text-gray-200
            hover:shadow dark:shadow-gray-700
            hover:bg-orange-100 dark:hover:bg-orange-700
            hover:rounded-lg
            "
      style="cursor: pointer;"
    >
      <span
        class="text-orange-600 dark:text-red-400"
      >
          
        1
        .
      </span>
       
      Bulbasaur
    </li>
  </ul>
</div>
`
